temporalio/worker/_workflow.py

```diff
@@ -191,15 +201,21 @@ async def _handle_activation(
                 logger.debug("Received workflow activation:\n%s", act)
 
             # We only have to run if there are any non-remove-from-cache jobs
-            remove_job = next(
-                (j for j in act.jobs if j.HasField("remove_from_cache")), None
-            )
-            if len(act.jobs) > 1 or not remove_job:
+            if not cache_remove_only_activation:
                 # If the workflow is not running yet, create it
                 workflow = self._running_workflows.get(act.run_id)
                 if not workflow:
-                    workflow = self._create_workflow_instance(act)
+                    # Must have a start job to create instance
+                    if not start_job:
+                        raise RuntimeError(
+                            "Missing start workflow, workflow could have unexpectedly been removed from cache"
+                        )
+                    workflow = self._create_workflow_instance(act, start_job)
                     self._running_workflows[act.run_id] = workflow
+                elif start_job:
+                    # This should never happen
+                    logger.warn("Cache already exists for activation with start job")
```

Comment:
If you have a thing like `dbg_assert!` that will throw during tests but not during prod, I would definitely use that here.

Reply:
There isn't a good one. There is `assert` which is based `__debug__` being `True` which it is by default (have to opt out with `-O`). However as these kinds of options grow, I could see a "strict mode" that we enable for tests.

Reply:
In that case I would at least run the tests once with this as an exception and make sure it's not happening.

If it doesn't take much time, I'd see about adding a way to get that behavior. It's pretty valuable and has saved me more than once on things like this.

Reply:
FWIW, I did a quick try on adding this same check in TS, and no test fail either.

Reply:
Also just switched to error and ran full suite locally with no issue here